design/design.md

```diff
@@ -0,0 +1,236 @@
+# Design for Streamly Coreutils package
+
+* cp
+
+  ```
+      data CpOptions = CpOptions {
+         attributesOnly :: Bool,
+         interactive :: Bool,
+         parents :: Bool,
+         symbolicLink :: Bool,
+         noTargetDir :: Bool,
+         verbose :: Bool,
+         noTargetDir :: Bool
+         ...
+      }
+
+      cpFile :: CpOptions -> SomeBase File -> SomeBase Dir -> SomeBase File -> IO ()
+
+  ```
+
+  1. Similar to srcFile -> destDir -> destFile
+  2. Copies the source file to a dest directory with dest file name
+  3. If dest file name is empty, then it becomes File -> Dir case (with same file name as source)
+
+  4. For File -> File within same directory, current directory should be passed as destDir argument
+   Else we can also say that an empty Dest Dir argument is equivalent to copying in the same directory
+
+  ```
+      cpFiletoDir :: (IsStream t, Monad m) => CpOptions -> t m (SomeBase File) -> SomeBase Dir -> IO ()
+         uses cpFile
+         (stream of input files so that copying can be done concurrently with use of appropriate streams)
+         or map can be used as well
+
+      cpDirToDir :: (IsStream t, Monad m) => CpOptions -> SomeBase Dir -> SomeBase Dir -> IO ()
+         should create a stream of files (t m (SomeBase File)) and call cpFileToDir
+
+  ```
+* echo
+
+  ```
+      data EchoOptions = CpOptions {
+         trailingLine :: Bool,
+         interpretBackSlash :: Bool,
+         ...
+      }
+
+      echo :: (IsStream t, Monad m) => EchoOptions -> t m Char -> SomeBase File -> IO ()
+
+  ```
+
+   1. for stdout, second arg should be "/dev/stdout"
+   2. writes the characters in the stream to the file
+   (as echo's output can be redirected to a file)
+
+* cat
+   1. concatenates files to standard output
+   2. [ if no file -> reads from std input ]
+
+  ```
+      data CatOptions = CpOptions {
+         showAll :: Bool,
+         numberNonEmptyLines :: Bool,
+         showEnds :: Bool,
+         numberAllLines :: Bool,
+         suppressRepeatedEmpty :: Bool,
+         ...
+      }
+
+      cat :: (IsStream t, Monad m) => CatOptions -> t m (SomeBase File) -> SomeBase File -> IO ()
+         for stdout, second arg should be "/dev/stdout"
+         reads the files, concatenates the output and writes in the destination file
+
+         if empty stream, accept input indefinitely
+
+         ( can be made streaming by not concatenating the output ; reading and writing parallely )
+
+      TODO: Redirection
+      Ex: cat *.md | sort
+
+  ```
+
+* wc
+      The wc utility code can be taken from the examples directory, WordCount.hs
+
+* yes
+
+  ```
+       yes :: String -> IO ()
```

Comment:
Why `String -> IO ()`? Will you just `print` it?
Maybe `String -> Stream m String`? Which is basically `S.repeat`

Reply:
:+1: